atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolImpl.java

```diff
@@ -324,24 +336,92 @@ private synchronized void refreshPool() {
 
     @VisibleForTesting
     void setServersInPoolTo(ImmutableSet<CassandraServer> desiredServers) {
-        ImmutableSet<CassandraServer> cachedServers = getCachedServers();
-        SetView<CassandraServer> serversToAdd = Sets.difference(desiredServers, cachedServers);
-        SetView<CassandraServer> absentServers = Sets.difference(cachedServers, desiredServers);
+        Set<CassandraServer> currentServers = getCachedServers();
+        SetView<CassandraServer> serversToAdd = Sets.difference(desiredServers, currentServers);
+        SetView<CassandraServer> absentServers = Sets.difference(currentServers, desiredServers);
 
-        serversToAdd.forEach(server -> cassandra.returnOrCreatePool(server, absentHostTracker.returnPool(server)));
         absentServers.forEach(cassandraServer -> {
             CassandraClientPoolingContainer container = cassandra.removePool(cassandraServer);
             absentHostTracker.trackAbsentCassandraServer(cassandraServer, container);
         });
 
         Set<CassandraServer> serversToShutdown = absentHostTracker.incrementAbsenceAndRemove();
 
-        if (!(serversToAdd.isEmpty() && absentServers.isEmpty())) { // if we made any changes
-            sanityCheckRingConsistency();
+        Set<CassandraServer> validatedServersToAdd =
+                validateNewHostsTopologiesAndMaybeAddToPool(getCurrentPools(), serversToAdd);
+
+        if (!(validatedServersToAdd.isEmpty() && absentServers.isEmpty())) { // if we made any changes
             cassandra.refreshTokenRangesAndGetServers();
         }
 
-        logRefreshedHosts(serversToAdd, serversToShutdown, absentServers);
+        Preconditions.checkState(
+                !getCurrentPools().isEmpty() || serversToAdd.isEmpty(),
+                "No servers were successfully added to the pool. This means we could not come to a consensus on"
+                    + " cluster topology, and the client cannot start as there are no valid hosts. This state should"
+                    + " be transient (<5 minutes), and if it is not, indicates that the user may have accidentally"
+                    + " configured AltasDB to use two separate Cassandra clusters (i.e., user-led split brain).",
```

Comment:
Can this happen _after_ startup? This refresh task will happen by default every 2 minutes, and due to runtime config, possibly(?) also after any runtime config changes. As such, this error message might want to change to reflect that (what happens if you fail to refresh your hosts but you've already started up?)

Reply:
yup makes sense, changed `start` to `connect`, although I suspect the case that this happens live to be extremely rare.